### DISCUSSION TITLE
Questions about the action 'InitialRandomSpaw'

### DISCUSSION THREAD

**Discussion by liuxz-cs** [allenai/ai2thor#743] [Q&A] [answered]
Excuse me, I have some question about the parameter of the action `InitialRandomSpaw`. In the parameter `numDuplicatesOfType`, we wet the count of each objects.  And the count is the max number of the object of the specific type because of the spawn failures sometimes. So if we set the cound bigger than 0, is it likely that the number of the object in the type is 0?  If so, how can we avoid that the actual number of a object type is 0? 
 
And what's the relationship between the parameter `numPlacementAttempts` and the count? Does the `numPlacementAttempts` mean the number  of the attempts of spawn of  each object of the specific type?  So, actually, the platform trys to spawn a object type for totally `count * numPlacementAttempts` times？

**Comment by winthos** [comment, marked as answer]
The `numDuplicatesOfType` is the number of attempted duplicates of an object that currently exists in the scene. It cannot be used to spawn in new objects that are not already present in the scene, as it makes a copy of an object that already exists. Because of this, you will never have the total number of objects be 0 for a specific object, because this parameter will only work if the original object(s) already exist by default in a scene.

Additionally, `numPlacementAttempts` is not at all related to the number of duplicates of an object specified by `numDuplicatesOfType`. What happens is, you pick an object to duplicate, and for each duplicate up to the `numDuplicatesOfType`, it will try and place that copy of the object on some free receptacle surface in the room. This is where `numPlacementAttempts` comes in. The higher the `numPlacementAttempts` is, the more spawn points on a receptacle will be tried to see if an object or a copy of an object will fit. When an object is shuffled by `InitialRandomSpawn`, what happens is a number of potential points within/on a receptacle are tried to see if an object can fit without colliding with anything. The `numPlacementAttempts` limits how many of these potential positions are checked, as checking all of the available positions will take a long time. This is an expensive collision check, so increasing the `numPlacementAttempts` will likely result in better success of moving an object to a new position, but it will cost performance.

To summarize, `numDuplicatesOfType` will be the maximum number of duplicates of an existing object that could exist in a scene. Higher `numPlacementAttempts` values allows greater accuracy to find a free position on a receptacle for either placing new duplicates specified by `numDuplicatesOfType`, or for just shuffling the positions of existing objects, at the cost of performance.

**Reply by liuxz-cs** [reply]
> The `numDuplicatesOfType` is the number of attempted duplicates of an object that currently exists in the scene. It cannot be used to spawn in new objects that are not already present in the scene, as it makes a copy of an object that already exists. Because of this, you will never have the total number of objects be 0 for a specific object, because this parameter will only work if the original object(s) already exist by default in a scene.
> 
> Additionally, `numPlacementAttempts` is not at all related to the number of duplicates of an object specified by `numDuplicatesOfType`. What happens is, you pick an object to duplicate, and for each duplicate up to the `numDuplicatesOfType`, it will try and place that copy of the object on some free receptacle surface in the room. This is where `numPlacementAttempts` comes in. The higher the `numPlacementAttempts` is, the more spawn points on a receptacle will be tried to see if an object or a copy of an object will fit. When an object is shuffled by `InitialRandomSpawn`, what happens is a number of potential points within/on a receptacle are tried to see if an object can fit without colliding with anything. The `numPlacementAttempts` limits how many of these potential positions are checked, as checking all of the available positions will take a long time. This is an expensive collision check, so increasing the `numPlacementAttempts` will likely result in better success of moving an object to a new position, but it will cost performance.
> 
> To summarize, `numDuplicatesOfType` will be the maximum number of duplicates of an existing object that could exist in a scene. Higher `numPlacementAttempts` values allows greater accuracy to find a free position on a receptacle for either placing new duplicates specified by `numDuplicatesOfType`, or for just shuffling the positions of existing objects, at the cost of performance.

Okay, so only if there already exists an `Apple` can I use `InitialRandomSpawn` and `numDuplicatesOfType` to get the duplicates of `Apple`. Then, if I set the `count` to be 2 in `numDuplicatesOfType` and  all the spawn attempts are successful, what's the number of `Apple`, 2 or 3? I want to ask about the mearning of `count`, does it mean the total number or the duplicate number except the one that already exists?

**Reply by winthos** [reply]
Ah I see. The `numDuplicatesOfType` sets the max total number of objects including the original. So for your example, if you set the `numDuplicatesOfType` of an apple that originally exists in the scene to `2`, you will end up with at most 2 total apples, one duplicate and the original apple. This is a little unintuitive, so we will add a warning in our documentation to make this more clear.

**Reply by liuxz-cs** [reply]
> Ah I see. The `numDuplicatesOfType` sets the max total number of objects including the original. So for your example, if you set the `numDuplicatesOfType` of an apple that originally exists in the scene to `2`, you will end up with at most 2 total apples, one duplicate and the original apple. This is a little unintuitive, so we will add a warning in our documentation to make this more clear.

Okay, thanks a lot!
